deps/fdasrsf/dp_grid.c

```diff
@@ -0,0 +1,257 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#include "dp_nbhd.h"
+#include "dp_grid.h"
+
+
+void dp_all_edge_weights(
+    double* Q1, double* T1, int nsamps1,
+    double* Q2, double* T2, int nsamps2,
+    int dim,
+    double* tv1, int* idxv1, int ntv1,
+    double* tv2, int* idxv2, int ntv2,
+    double* W, double lam)
+{
+    int sr, sc;  /* source row and column */
+    int tr, tc;  /* target row and column */
+    int l1, l2, l3;  /* for multidimensional array mapping */
+    int i;
+
+    for (i=0; i<ntv1*ntv2*ntv1*ntv2; W[i++]=1e6);
+
+    /* W is a ntv2 x ntv1 x ntv2 x ntv1 array.
+     * Weight of edge from (tv1[i],tv2[j]) to (tv1[k],tv2[l])
+     * (Cartesian coordinates) is in grid(j,i,l,k).
+     * Mapping:
+     *  (j,i,l,k) :--> j*ntv1*ntv2*ntv1 +
+     *                 i*ntv2*ntv1 +
+     *                 l*ntv1 +
+     *                 k
+     */
+    l1 = ntv1 * ntv2 * ntv1;
+    l2 = ntv2 * ntv1;
+    l3 = ntv1;
+
+    for (tr=1; tr<ntv2; ++tr) {
+        for (tc=1; tc<ntv1; ++tc) {
+            for (i=0; i<DP_NBHD_COUNT; ++i) {
+                sr = tr - dp_nbhd[i][0];
+                sc = tc - dp_nbhd[i][1];
+
+                if (sr < 0 || sc < 0) continue;
+
+                /* grid(sr,sc,tr,tc) */
+                W[sr*l1+sc*l2+tr*l3+tc] =
+                    dp_edge_weight(Q1, T1, nsamps1, Q2, T2, nsamps2, dim,
+                                   tv1[sc], tv1[tc], tv2[sr], tv2[tr],
+                                   idxv1[sc], idxv2[sr], lam);
+            }
+        }
+    }
+}
+
+
+double dp_costs(
+    double* Q1, double* T1, int nsamps1,
+    double* Q2, double* T2, int nsamps2,
+    int dim,
+    double* tv1, int* idxv1, int ntv1,
+    double* tv2, int* idxv2, int ntv2,
+    double* E, int* P, double lam)
+{
+    int sr, sc;  /* source row and column */
+    int tr, tc;  /* target row and column */
+    double w, cand_cost;
+    int i;
+
+    E[0] = 0.0;
+    for (i=1; i<ntv1; E[i++]=1e6);
```

Comment:
Why so many `for` for the initialization of the same array?

Reply:
``E`` is a 2-D array where it is computed the energy function. It should be initialised with E[0, 1:]= ∞ and E[1:, 0]=∞ and E[0,0]=0.